File: AngularJS/src/org/angular2/entities/Angular2DirectiveSelectorImpl.java
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.entities;

import com.intellij.openapi.util.AtomicNotNullLazyValue;
import com.intellij.openapi.util.ClearableLazyValue;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.ElementManipulators;
import com.intellij.psi.PsiElement;
import com.intellij.util.SmartList;
import com.intellij.util.containers.JBIterable;
import org.angular2.lang.selector.Angular2DirectiveSimpleSelector;
import org.angular2.lang.selector.Angular2DirectiveSimpleSelector.Angular2DirectiveSimpleSelectorWithRanges;
import org.angular2.lang.selector.Angular2DirectiveSimpleSelector.ParseException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;
import java.util.function.Supplier;

public class Angular2DirectiveSelectorImpl implements Angular2DirectiveSelector {

  private final ClearableLazyValue<PsiElement> myLazyParent;
  private PsiElement myStaticParent;
  private final String myText;
  private final Function<? super Pair<String, Integer>, ? extends TextRange> myCreateRange;
  private final AtomicNotNullLazyValue<List<Angular2DirectiveSimpleSelector>> mySimpleSelectors =
    new AtomicNotNullLazyValue<List<Angular2DirectiveSimpleSelector>>() {
      @Override
      protected @NotNull List<Angular2DirectiveSimpleSelector> compute() {
        if (myText == null) {
          return Collections.emptyList();
        }
        try {
          return Collections.unmodifiableList(Angular2DirectiveSimpleSelector.parse(myText));
        }
        catch (ParseException e) {
          return Collections.emptyList();
        }
      }
    };
  private final AtomicNotNullLazyValue<List<SimpleSelectorWithPsi>> mySimpleSelectorsWithPsi =
    new AtomicNotNullLazyValue<List<SimpleSelectorWithPsi>>() {
      @Override
      protected @NotNull List<SimpleSelectorWithPsi> compute() {
        if (myText == null) {
          return Collections.emptyList();
        }
        try {
          List<Angular2DirectiveSimpleSelectorWithRanges> simpleSelectorsWithRanges = Angular2DirectiveSimpleSelector.parseRanges(myText);
          List<SimpleSelectorWithPsi> result = new ArrayList<>(simpleSelectorsWithRanges.size());
          for (Angular2DirectiveSimpleSelectorWithRanges sel : simpleSelectorsWithRanges) {
            result.add(new SimpleSelectorWithPsiImpl(sel));
          }
          return Collections.unmodifiableList(result);
        }
        catch (ParseException e) {
          return Collections.emptyList();
        }
      }
    };

  public Angular2DirectiveSelectorImpl(@NotNull PsiElement element,
                                       @Nullable String text,
                                       @Nullable Function<? super Pair<String, Integer>, ? extends TextRange> createRange) {
    myLazyParent = null;
    myStaticParent = element;
    myText = text;
    myCreateRange = createRange != null ? createRange : a -> TextRange.EMPTY_RANGE;
  }

  public Angular2DirectiveSelectorImpl(@NotNull Supplier<? extends PsiElement> element,
                                       @Nullable String text,
                                       @Nullable Function<? super Pair<String, Integer>, ? extends TextRange> createRange) {
    myLazyParent = ClearableLazyValue.createAtomic(element);
    myText = text;
    myCreateRange = createRange != null ? createRange : a -> TextRange.EMPTY_RANGE;
  }

  @Override
  public @NotNull String getText() {
    return myText == null ? "<null>" : myText;
  }

  public PsiElement getPsiParent() {
    return myStaticParent != null ? myStaticParent : myLazyParent != null ? myLazyParent.getValue() : null;
  }

  @Override
  public @NotNull List<Angular2DirectiveSimpleSelector> getSimpleSelectors() {
    return mySimpleSelectors.getValue();
  }

  @Override
  public @NotNull List<SimpleSelectorWithPsi> getSimpleSelectorsWithPsi() {
    return mySimpleSelectorsWithPsi.getValue();
  }

  @Override
  public @NotNull Angular2DirectiveSelectorPsiElement getPsiElementForElement(@NotNull String elementName) {
    for (SimpleSelectorWithPsi selector : getSimpleSelectorsWithPsi()) {
      if (selector.getElement() != null && elementName.equalsIgnoreCase(selector.getElement().getName())) {
        return selector.getElement();
      }
      for (SimpleSelectorWithPsi notSelector : selector.getNotSelectors()) {
        if (notSelector.getElement() != null && elementName.equalsIgnoreCase(notSelector.getElement().getName())) {
          return notSelector.getElement();
        }
      }
    }
    return new Angular2DirectiveSelectorPsiElement(this, new TextRange(0, 0), elementName, true);
  }

  @Override
  public String toString() {
    return getText();
  }

  protected @NotNull Angular2DirectiveSelectorPsiElement convert(@NotNull Pair<String, Integer> range, boolean isElement) {
    return new Angular2DirectiveSelectorPsiElement(this, myCreateRange.apply(range), range.first, isElement);
  }

  public void replaceText(@NotNull TextRange range, @NotNull String name) {
    myStaticParent = ElementManipulators.getManipulator(myStaticParent)
      .handleContentChange(myStaticParent, range, name);
  }

  private class SimpleSelectorWithPsiImpl implements SimpleSelectorWithPsi {

    private final Angular2DirectiveSelectorPsiElement myElement;
    private final List<Angular2DirectiveSelectorPsiElement> myAttributes = new SmartList<>();
    private final List<SimpleSelectorWithPsi> myNotSelectors = new SmartList<>();

    SimpleSelectorWithPsiImpl(@NotNull Angular2DirectiveSimpleSelectorWithRanges selectorWithRanges) {
      if (selectorWithRanges.getElementRange() != null) {
        myElement = convert(selectorWithRanges.getElementRange(), true);
      }
      else {
        myElement = null;
      }
      for (Pair<String, Integer> attr : selectorWithRanges.getAttributeRanges()) {
        myAttributes.add(convert(attr, false));
      }
      for (Angular2DirectiveSimpleSelectorWithRanges notSelector : selectorWithRanges.getNotSelectors()) {
        myNotSelectors.add(new SimpleSelectorWithPsiImpl(notSelector));
      }
    }

    @Override
    public @Nullable Angular2DirectiveSelectorPsiElement getElement() {
      return myElement;
    }

    @Override
    public @NotNull List<@NotNull Angular2DirectiveSelectorPsiElement> getAttributes() {
      return myAttributes;
    }

    @Override
    public @NotNull List<@NotNull SimpleSelectorWithPsi> getNotSelectors() {
      return myNotSelectors;
    }

    @Override
    public @Nullable Angular2DirectiveSelectorPsiElement getElementAt(int offset) {
      return JBIterable.from(myAttributes)
        .append(JBIterable.from(myNotSelectors).flatMap(sel -> sel.getAttributes()))
        .append(myElement)
        .filter(element -> element.getTextRangeInParent().contains(offset))
        .first();
    }
  }
}

File: AngularJS/src/org/angular2/entities/Angular2DirectiveSelectorImpl.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.entities

import com.intellij.model.Pointer
import com.intellij.openapi.util.NotNullLazyValue
import com.intellij.openapi.util.Pair
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.refactoring.suggested.createSmartPointer
import com.intellij.util.SmartList
import com.intellij.util.concurrency.SynchronizedClearableLazy
import org.angular2.entities.Angular2DirectiveSelector.SimpleSelectorWithPsi
import org.angular2.entities.metadata.psi.Angular2MetadataDirectiveBase
import org.angular2.lang.selector.Angular2DirectiveSimpleSelector
import org.angular2.lang.selector.Angular2DirectiveSimpleSelector.Angular2DirectiveSimpleSelectorWithRanges
import org.angular2.lang.selector.Angular2DirectiveSimpleSelector.ParseException
import java.util.*

class Angular2DirectiveSelectorImpl(private val myElement: PsiElement,
                                    private val myText: String?,
                                    private val myRangeOffset: Int?) : Angular2DirectiveSelector {
  private val myLazyParent: SynchronizedClearableLazy<PsiElement>? = if (myElement is Angular2MetadataDirectiveBase<*>)
    SynchronizedClearableLazy { myElement.typeScriptClass ?: myElement }
  else
    null
  private val mySimpleSelectors: NotNullLazyValue<List<Angular2DirectiveSimpleSelector>> = NotNullLazyValue.lazy {
    if (myText == null) {
      emptyList()
    }
    else try {
      Collections.unmodifiableList(Angular2DirectiveSimpleSelector.parse(myText))
    }
    catch (e: ParseException) {
      emptyList()
    }
  }
  private val mySimpleSelectorsWithPsi: NotNullLazyValue<List<SimpleSelectorWithPsi>> = NotNullLazyValue.lazy {
    if (myText == null) {
      emptyList()
    }
    else try {
      Angular2DirectiveSimpleSelector.parseRanges(myText)
        .map { SimpleSelectorWithPsiImpl(it, null) }
    }
    catch (e: ParseException) {
      emptyList()
    }
  }

  override val text: String
    get() = myText ?: "<null>"

  val psiParent: PsiElement
    get() = if (myElement is Angular2MetadataDirectiveBase<*>) myLazyParent!!.value else myElement

  override val simpleSelectors: List<Angular2DirectiveSimpleSelector>
    get() = mySimpleSelectors.value

  override val simpleSelectorsWithPsi: List<SimpleSelectorWithPsi>
    get() = mySimpleSelectorsWithPsi.value

  fun createPointer(): Pointer<Angular2DirectiveSelectorImpl> {
    val element = myElement.createSmartPointer()
    val text = myText
    val rangeOffset = myRangeOffset
    return Pointer {
      element.element?.let { Angular2DirectiveSelectorImpl(it, text, rangeOffset) }
    }
  }

  override fun getSymbolForElement(elementName: String): Angular2DirectiveSelectorSymbol {
    for (selector in simpleSelectorsWithPsi) {
      val selectorElement = selector.element
      if (selectorElement != null && elementName.equals(selectorElement.name, ignoreCase = true)) {
        return selectorElement
      }
      for (notSelector in selector.notSelectors) {
        val notSelectorElement = notSelector.element
        if (notSelectorElement != null && elementName.equals(notSelectorElement.name, ignoreCase = true)) {
          return notSelectorElement
        }
      }
    }
    return Angular2DirectiveSelectorSymbol(this, TextRange(0, 0), elementName, null, true)
  }

  override fun toString(): String {
    return text
  }

  private fun convert(range: Pair<String, Int>,
                      elementSelector: String?,
                      isElement: Boolean): Angular2DirectiveSelectorSymbol {
    return Angular2DirectiveSelectorSymbol(
      this,
      if (myRangeOffset != null)
        TextRange(range.second + myRangeOffset, range.second + range.first.length + myRangeOffset)
      else
        TextRange.EMPTY_RANGE,
      range.first, elementSelector, isElement)
  }

  /*fun replaceText(range: TextRange, name: String) {
    myElement = ElementManipulators.getManipulator(myElement!!)!!
      .handleContentChange(myElement!!, range, name)
  }*/

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val selector = other as Angular2DirectiveSelectorImpl?
    return myElement == selector!!.myElement &&
           myText == selector.myText &&
           myRangeOffset == selector.myRangeOffset
  }

  override fun hashCode(): Int {
    return Objects.hash(myElement, myText, myRangeOffset)
  }

  private inner class SimpleSelectorWithPsiImpl(selectorWithRanges: Angular2DirectiveSimpleSelectorWithRanges,
                                                mainElementSelector: String?) : SimpleSelectorWithPsi {

    override val element: Angular2DirectiveSelectorSymbol?
    private val myAttributes = SmartList<Angular2DirectiveSelectorSymbol>()
    private val myNotSelectors = SmartList<SimpleSelectorWithPsi>()

    override val attributes: List<Angular2DirectiveSelectorSymbol>
      get() = myAttributes

    override val notSelectors: List<SimpleSelectorWithPsi>
      get() = myNotSelectors

    init {
      var myElementName: String? = null
      if (selectorWithRanges.elementRange != null) {
        element = convert(selectorWithRanges.elementRange!!, null, true)
        myElementName = element.name
      }
      else {
        element = null
      }
      for (attr in selectorWithRanges.attributeRanges) {
        myAttributes.add(convert(attr, myElementName ?: mainElementSelector, false))
      }
      for (notSelector in selectorWithRanges.notSelectors) {
        myNotSelectors.add(SimpleSelectorWithPsiImpl(notSelector, myElementName))
      }
    }

    override fun getElementAt(offset: Int): Angular2DirectiveSelectorSymbol? {
      return myAttributes
        .plus(myNotSelectors.flatMap { it.attributes })
        .plus(element)
        .firstOrNull { it?.textRangeInSource?.contains(offset) == true }
    }
  }
}
